tests/builder/test_base.py

```diff
@@ -125,3 +159,88 @@ def count_markers(path, marker='# Edited by rpmlb'):
         edit_file(empty_spec_path, indicator)
         assert empty_spec_path.exists()
         assert count_markers(empty_spec_path) == 1
+
+
+def test_macros_are_added_correctly(macro_spec_path, prepared_macros):
+    """Macros are safely added without messing with current file contents."""
+
+    new_macro_dict = {
+        'bootstrap': 1,
+        'with_bootstrap': 1,
+    }
+
+    with BaseBuilder.edit_spec_file(macro_spec_path) as (original, modified):
+        content_stream = BaseBuilder.add_macros(original, new_macro_dict)
+        modified.write(''.join(content_stream))
+
+    with macro_spec_path.open() as spec_file:
+        spec_contents = spec_file.read()
+
+    # Count the macro occurrences
+    matches = MACRO_REGEX.finditer(spec_contents)
+    macro_counts = Counter(match.group('name') for match in matches)
+
+    # Each expected macro is present exactly once
+    assert all(cnt == 1 for cnt in macro_counts.values()), macro_counts
+
+    expected_macro_names = new_macro_dict.keys() | prepared_macros.keys()
+    present_macro_names = macro_counts.keys()
+    assert expected_macro_names == present_macro_names, present_macro_names
```

Comment:
I'm quite confident pytest would tell you why the `==` does not apply and what is the content of both `expected_macro_names` and `present_macro_names`, so adding it as the second assert argument is not necessary (but it doesn't harm) (applies to multiple locations)